File: spikes/boltffi-retest/host/Contract.kt
package dev.greenfield5.boltproof

import dev.greenfield5.bolt.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*

// Called from instrumentation: every operation below goes through the generated
// BoltFFI Kotlin bindings into real JNI into the real Rust library. Nothing
// here re-implements product behaviour in Kotlin.
object Contract {
    fun run() = runBlocking {
        check(coreVersion() == "0.1.0")
        GreenfieldSession(Role.SENDER, ConnectionMode.INTERNET).use { sender ->
            check(sender.stateCode() == 0.toUByte())
            check(sender.apply(SessionCommand.START_PAIRING) == SessionState.AWAITING_PEER)
            check(sender.sendCommand(2u) == 2.toUByte())
            try { sender.sendCommand(2u); error("admitted second viewer") }
            catch (e: BridgeError.ViewerAlreadyConnected) { }
            check(sender.sendCommand(3u) == 3.toUByte())
            check(sender.viewerApproved())
            try { sender.sendCommand(255u); error("accepted invalid command") }
            catch (e: BridgeError.UnknownCommandCode) { check(e.code == 255.toUByte()) }
            check(sender.stateCode() == 3.toUByte())
        }
        GreenfieldSession(Role.VIEWER, ConnectionMode.LOCAL).use { viewer ->
            check(viewer.sendCommand(1u) == 2.toUByte())
            check(viewer.sendCommand(4u) == 3.toUByte())
        }
        try { GreenfieldSession.fromCodes(255u, 0u); error("accepted invalid role") }
        catch (e: BridgeError.UnknownRoleCode) { check(e.code == 255.toUByte()) }
        try { GreenfieldSession.fromCodes(0u, 255u); error("accepted invalid mode") }
        catch (e: BridgeError.UnknownModeCode) { check(e.code == 255.toUByte()) }
        val padded = PaddingProbe(0x1234u, 0x55667788u)
        check(paddingRoundTrip(padded) == padded)
        check(paddingList(padded) == listOf(padded, padded))
        val input = LayoutProbe(7u, 0x1122334455667788uL, 0x3344u, "layout")
        check(roundTrip(input) == input)

        // async success / typed async failure / cleanup after the future settles
        AsyncProbe().use { probe ->
            probe.release()
            check(probe.waitValue(42u, false) == 42u)
            try { probe.waitValue(42u, true); error("missing typed error") }
            catch (e: BridgeError.SessionEnded) { }
            check(probe.active() == 0u)
        }

        // cancellation before completion; cleanup after cancellation
        AsyncProbe().use { probe ->
            val task = launch(Dispatchers.Default) { probe.waitValue(1u, false) }
            withTimeout(3000) { while (probe.active() == 0u) yield() }
            task.cancelAndJoin()
            check(probe.active() == 0u)
        }

        // repeated cancellation: cancel twice, and once after the future settled.
        // The generated runtime must free the native future exactly once.
        var repeatedCancellations = 0
        repeat(100) {
            AsyncProbe().use { probe ->
                val task = launch(Dispatchers.Default) { runCatching { probe.waitValue(1u, false) } }
                withTimeout(3000) { while (probe.active() == 0u) yield() }
                task.cancel()
                task.cancel()
                withTimeout(3000) { task.join() }
                check(probe.active() == 0u)
                repeatedCancellations++
            }
        }

        // cancellation racing readiness: release() makes the future ready while
        // the cancel lands. One of the two paths must claim the completion.
        var racedCancellations = 0
        repeat(100) {
            AsyncProbe().use { probe ->
                val task = launch(Dispatchers.Default) { runCatching { probe.waitValue(5u, false) } }
                withTimeout(3000) { while (probe.active() == 0u) yield() }
                probe.release()
                task.cancel()
                withTimeout(3000) { task.join() }
                check(probe.active() == 0u)
                racedCancellations++
            }
        }

        // bounded batch path: native ring capacity is 8, producer overshoots it
        EventProbe().use { probe ->
            probe.eventsBatch().use { stream ->
                check(probe.produce(100u) == 8u)
                val batch = stream.popBatch(100)
                check(batch == (0u..7u).toList())
                check(probe.dropped() == 92u)
                probe.stop()
                stream.unsubscribe()
            }
        }

        // normal asynchronous consumer: every accepted event arrives, in order
        EventProbe().use { probe ->
            val seen = mutableListOf<UInt>()
            val consumer = launch { probe.events().take(8).collect { seen.add(it) } }
            yield()
            check(probe.produce(8u) == 8u)
            withTimeout(3000) { consumer.join() }
            check(seen == (0u..7u).toList())
            check(probe.dropped() == 0u)
            check(!probe.isActive())
        }

        // slow consumer: produce one at a time so the native ring never fills,
        // consume slowly. Records the resulting host backlog explicitly.
        EventProbe().use { probe ->
            val seen = mutableListOf<UInt>()
            val consumer = launch { probe.events().collect { seen.add(it); delay(10) } }
            yield()
            repeat(200) { probe.produce(1u); delay(1) }
            consumer.cancelAndJoin()
            check(seen.zipWithNext().all { (a, b) -> a < b })
            check(probe.produced() == 200u)
            check(!probe.isActive())
            println(
                "BOLT_STREAM produced=${probe.produced()} consumed=${seen.size} " +
                    "nativeDropped=${probe.dropped()} unconsumed=${200 - seen.size - probe.dropped().toInt()}"
            )
        }

        // shutdown: drain, stop the producer, then confirm the subscription is gone
        EventProbe().use { probe ->
            probe.eventsBatch().use { stream ->
                probe.produce(3u)
                check(stream.popBatch(8).size == 3)
                check(stream.popBatch(8).isEmpty())
                probe.stop()
                check(!probe.isActive())
                stream.unsubscribe()
            }
        }

        println(
            "BOLT_PROOF version=0.1.0 sender=PASS viewer=PASS typed_errors=PASS layout=PASS " +
                "async=PASS cancellation=PASS repeated_cancel=$repeatedCancellations " +
                "raced_cancel=$racedCancellations batch=8/100 dropped=92"
        )
    }
}

File: spikes/boltffi-retest/tests/contract.rs
use greenfield5_bolt_spike::*;

#[test]
fn native_core_version_is_exact() {
    assert_eq!(core_version(), "0.1.0");
}

#[test]
fn sender_viewer_and_rejection_contract() {
    let sender = GreenfieldSession::new(Role::Sender, ConnectionMode::Internet);
    assert_eq!(sender.state_code(), 0);
    assert_eq!(sender.apply(SessionCommand::StartPairing).unwrap(), SessionState::AwaitingPeer);
    assert_eq!(sender.send_command(2).unwrap(), 2);
    assert_eq!(sender.send_command(2), Err(BridgeError::ViewerAlreadyConnected));
    assert_eq!(sender.send_command(3).unwrap(), 3);
    assert!(sender.viewer_approved());
    assert_eq!(sender.send_command(255), Err(BridgeError::UnknownCommandCode { code: 255 }));
    assert_eq!(sender.state_code(), 3);
    let viewer = GreenfieldSession::from_codes(1, 0).unwrap();
    assert_eq!(viewer.send_command(1).unwrap(), 2);
    assert_eq!(viewer.send_command(4).unwrap(), 3);
    assert!(matches!(GreenfieldSession::from_codes(255, 0), Err(BridgeError::UnknownRoleCode { code: 255 })));
    assert!(matches!(GreenfieldSession::from_codes(0, 255), Err(BridgeError::UnknownModeCode { code: 255 })));
}

#[test]
fn mixed_layout_round_trip() {
    let input = LayoutProbe { flag: 7, wide: 0x1122334455667788, short: 0x3344, text: "layout".into() };
    assert_eq!(round_trip(input.clone()), input);
}

#[test]
fn native_stream_capacity_drop_and_stop() {
    let probe = EventProbe::new();
    let stream = probe.events_batch();
    assert_eq!(probe.produce(100), 8);
    assert_eq!(probe.produced(), 100);
    assert_eq!(probe.dropped(), 92);
    for sequence in 0..8 { assert_eq!(stream.pop_event(), Some(sequence)); }
    assert_eq!(stream.pop_event(), None);
    probe.stop();
    assert!(!stream.is_active());
}

#[test]
fn async_future_cancel_releases_guard() {
    use std::{future::Future, sync::Arc, task::{Context, Poll, Wake, Waker}};
    struct Noop;
    impl Wake for Noop { fn wake(self: Arc<Self>) {} }
    let waker = Waker::from(Arc::new(Noop));
    let mut context = Context::from_waker(&waker);
    let probe = AsyncProbe::new();
    let mut future = Box::pin(probe.wait_value(42, false));
    assert_eq!(future.as_mut().poll(&mut context), Poll::Pending);
    assert_eq!(probe.active(), 1);
    drop(future);
    assert_eq!(probe.active(), 0);
    probe.release();
    let mut success = Box::pin(probe.wait_value(42, false));
    assert_eq!(success.as_mut().poll(&mut context), Poll::Ready(Ok(42)));
    let mut failure = Box::pin(probe.wait_value(42, true));
    assert_eq!(failure.as_mut().poll(&mut context), Poll::Ready(Err(BridgeError::SessionEnded)));
    assert_eq!(probe.active(), 0);
}

#[test]
fn padding_struct_and_vector() {
    let value = PaddingProbe { version: 0x1234, value: 0x55667788 };
    assert_eq!(padding_round_trip(value), value);
    assert_eq!(padding_list(value), vec![value, value]);
}
